### PR TITLE
Libc: More libc functions for Redis(Part 3)

### PR DESCRIPTION
**Pull request by coolyjg**
Most of this pr are function declarations and "unimplemented" definitions, EXCEPT: 
- `rename` implementation.
- `pow` and `log` implementated by C.
- `sysfs` and `procfs` implemented by `ramfs`.
- `mktime` and `strftime` implemented by Rust, following relibc.

### CODE REVIEW

**Comment by equation314** [comment]
Please remove `ulib/libax`, `ulib/c_libax`

File: ulib/axlibc/c/file.c
Comment (equation314) [inline ulib/axlibc/c/file.c:1]: `file.c` -> `flock.c`

File: ulib/axlibc/c/netdb.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                           "\0Unknown error";
+
+const char *__lctrans_cur(const char *msg)
+{
+    return msg;
+}
+
+#define LCTRANS_CUR(msg) __lctrans_cur(msg)
+
+const char *gai_strerror(int ecode)
```
Comment (equation314) [inline ulib/axlibc/c/netdb.c]: Move to `network.c` near `getaddrinfo`/`freeaddrinfo`

File: ulib/axlibc/c/network.c [hunk 2]
```diff
@@ -71,6 +74,15 @@ static __inline uint32_t __bswap_32(uint32_t __x)
     return __x >> 24 | (__x >> 8 & 0xff00) | (__x << 8 & 0xff0000) | __x << 24;
 }
 
+static __inline uint64_t __bswap_64(uint64_t __x)
```
Comment (equation314) [inline ulib/axlibc/c/network.c]: Not used?

File: ulib/axlibc/src/file.rs [hunk 3]
```diff
@@ -192,3 +192,17 @@ pub unsafe extern "C" fn ax_getcwd(buf: *mut c_char, size: usize) -> *mut c_char
...
 }
+
+/// Rename `old` file to `new`
+#[no_mangle]
+pub unsafe extern "C" fn ax_rename(old: *const c_char, new: *const c_char) -> c_int {
+    ax_call_body!(ax_rename, {
+        let old_path = char_ptr_to_str(old)?;
+        let new_path = char_ptr_to_str(new)?;
+        debug!("ax_rename <= old: {:?}, new: {:?}", old_path, new_path);
+        match fs::rename(old_path, new_path) {
```
Comment (equation314) [inline ulib/axlibc/src/file.rs]: Returns `fs::rename(old_path, new_path)` directly

File: ulib/axlibc/src/mktime.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+];
+
+#[inline(always)]
+fn leap_year(year: c_int) -> bool {
+    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
+}
+
+/// `mktime` implementation
+#[no_mangle]
+pub unsafe extern "C" fn ax_mktime(t: *mut ctypes::tm) -> ctypes::time_t {
```
Comment (equation314) [inline ulib/axlibc/src/mktime.rs:19]: Call `ax_mktime` in C function `mktime`

File: ulib/axlibc/c/pwd.c [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    }
+
+done:
+    pthread_setcancelstate(cs, 0);
+    if (rv)
+        errno = rv;
+    return rv;
+}
+
+static int getpw_r(const char *name, uid_t uid, struct passwd *pw, char *buf, size_t size,
```
Comment (equation314) [inline ulib/axlibc/c/pwd.c]: Password-related functions are not planning to support in ArceOS, leave them unimplemented.

File: modules/axfs/src/api/mod.rs [hunk 6]
```diff
@@ -79,3 +79,8 @@ pub fn remove_dir(path: &str) -> io::Result<()> {
 pub fn remove_file(path: &str) -> io::Result<()> {
     crate::root::remove_file(None, path)
 }
+
+/// Renames a file
```
Comment (equation314) [inline modules/axfs/src/api/mod.rs]: Add more detailed comments like the rust std library, and add comments about the limitation.

File: modules/axfs/src/fops.rs [hunk 7]
```diff
@@ -328,6 +328,11 @@ impl Directory {
         self.entry_idx += n;
         Ok(n)
     }
+
+    /// Renames a file
+    pub fn rename(&self, old: &str, new: &str) -> AxResult {
```
Comment (equation314) [inline modules/axfs/src/fops.rs:336]: Add comments about the limitation.

File: modules/axfs/src/fs/fatfs.rs [hunk 8]
```diff
@@ -184,6 +184,17 @@ impl VfsNodeOps for DirWrapper<'static> {
         }
         Ok(dirents.len())
     }
+
+    fn rename(&self, src_path: &str, dst_path: &str) -> VfsResult {
+        debug!(
+            "rename at fatfs, src_path: {}, dst_path: {}",
+            src_path, dst_path
+        );
+
```
Comment (equation314) [inline modules/axfs/src/fs/fatfs.rs:194]: Add comments about the limitation.

File: modules/axfs/Cargo.toml [hunk 9]
```diff
@@ -15,8 +15,10 @@ ramfs = ["dep:axfs_ramfs"]
 fatfs = ["dep:fatfs"]
 myfs = ["dep:crate_interface"]
 use-ramdisk = []
+procfs = []
```
Comment (equation314) [inline modules/axfs/Cargo.toml]: More these two lines after `ramfs`, and add dependencies to `dep:axfs_ramfs`.

File: modules/axfs/src/root.rs [hunk 10]
```diff
@@ -131,6 +131,96 @@ impl VfsNodeOps for RootDirectory {
...
+            if rest_path.is_empty() {
+                ax_err!(NotFound)
+            } else {
+                fs.root_dir().rename(rest_path, dst_path)
+            }
+        })
+    }
+}
+
+#[cfg(all(feature = "ramfs", feature = "procfs"))]
```
Comment (equation314) [inline modules/axfs/src/root.rs]: Remove `feature = "ramfs"`

File: modules/axfs/src/root.rs [hunk 11]
```diff
@@ -131,6 +131,96 @@ impl VfsNodeOps for RootDirectory {
...
+fn mount_procfs() -> fs::ramfs::RamFileSystem {
```
Comment (equation314) [inline modules/axfs/src/root.rs]: Two many `unwrap`s, use the `?` expression.

File: modules/axfs/src/root.rs [hunk 12]
```diff
@@ -171,6 +261,18 @@ pub(crate) fn init_rootfs(disk: crate::dev::Disk) {
             .expect("failed to mount ramfs at /tmp");
     }
 
+    // Mount another ramfs as procfs
```
Comment (equation314) [inline modules/axfs/src/root.rs:170]: Move code pieces about `devfs` and `ramfs` as two separated functions.

**Review by equation314 — CHANGES_REQUESTED** [review]

File: ulib/axlibc/src/file.rs (hunk 3 above)
Comment (coolyjg) [reply ulib/axlibc/src/file.rs]: The error type does not match

File: ulib/axlibc/src/file.rs [hunk 13]
```diff
@@ -192,3 +192,17 @@ pub unsafe extern "C" fn ax_getcwd(buf: *mut c_char, size: usize) -> *mut c_char
...
+/// If new exists, it is first removed.
+///
+/// Return 0 if the operation succeeds, otherwise return -1.
+#[no_mangle]
+pub unsafe extern "C" fn ax_rename(old: *const c_char, new: *const c_char) -> c_int {
+    ax_call_body!(ax_rename, {
+        let old_path = char_ptr_to_str(old)?;
+        let new_path = char_ptr_to_str(new)?;
+        debug!("ax_rename <= old: {:?}, new: {:?}", old_path, new_path);
+        fs::rename(old_path, new_path).map_or_else(|e| Err(LinuxError::from(e)), |_| Ok(0))
```
Comment (equation314) [inline ulib/axlibc/src/file.rs]:
```rust
axstd::fs::rename(old_path, new_path)?;
Ok(0)
```

File: ulib/axlibc/c/network.c [hunk 14]
```diff
@@ -71,6 +106,9 @@ static __inline uint32_t __bswap_32(uint32_t __x)
     return __x >> 24 | (__x >> 8 & 0xff00) | (__x << 8 & 0xff0000) | __x << 24;
 }
 
+#define bswap_16(x) __bswap_16(x)
+#define bswap_32(x) __bswap_32(x)
```
Comment (equation314) [inline ulib/axlibc/c/network.c]: These two definitions are not used

File: ulib/axlibc/c/network.c [hunk 15]
```diff
@@ -44,6 +47,38 @@ void freeaddrinfo(struct addrinfo *__restrict res)
     return;
 }
 
+static const char msgs_lc[] = "Invalid flags\0"
```
Comment (equation314) [inline ulib/axlibc/c/network.c]: `gai_msgs`

File: ulib/axlibc/c/network.c [hunk 16]
```diff
@@ -44,6 +47,38 @@ void freeaddrinfo(struct addrinfo *__restrict res)
...
+
+const char *gai_strerror(int ecode)
+{
+    const char *s;
+    for (s = msgs_lc, ecode++; ecode && *s; ecode++, s++)
+        for (; *s; s++)
+            ;
+    if (!*s)
+        s++;
+    return LCTRANS_CUR(s);
```
Comment (equation314) [inline ulib/axlibc/c/network.c]: `return s;` is okay

File: modules/axfs/src/root.rs [hunk 17]
```diff
@@ -148,28 +234,26 @@ pub(crate) fn init_rootfs(disk: crate::dev::Disk) {
...
-        root_dir
-            .mount("/tmp", Arc::new(ramfs))
-            .expect("failed to mount ramfs at /tmp");
-    }
+    root_dir
+        .mount("/tmp", Arc::new(mount_ramfs()))
+        .expect("failed to mount ramfs at /tmp");
+
+    // Mount another ramfs as procfs
+    #[cfg(all(feature = "ramfs", feature = "procfs"))]
```
Comment (equation314) [inline modules/axfs/src/root.rs]: `#[cfg(feature = "procfs")]`

File: modules/axfs/src/root.rs [hunk 18]
```diff
@@ -148,28 +234,26 @@ pub(crate) fn init_rootfs(disk: crate::dev::Disk) {
...
+    root_dir // should not fail
+        .mount("/proc", Arc::new(mount_procfs().unwrap()))
+        .expect("fail to mount procfs at /proc");
+
+    // Mount another ramfs as sysfs
+    #[cfg(all(feature = "ramfs", feature = "sysfs"))]
+    root_dir // should not fail
```
Comment (equation314) [inline modules/axfs/src/root.rs:178]: `#[cfg(feature = "sysfs")]`

File: modules/axfs/src/root.rs [hunk 19]
```diff
@@ -131,6 +131,92 @@ impl VfsNodeOps for RootDirectory {
...
+                ax_err!(NotFound)
+            } else {
+                fs.root_dir().rename(rest_path, dst_path)
+            }
+        })
+    }
+}
+
+#[cfg(feature = "devfs")]
+fn mount_devfs() -> fs::devfs::DeviceFileSystem {
```
Comment (equation314) [inline modules/axfs/src/root.rs]: Returns `Arc<...>` for all `mount_*` functions.

Comment (equation314) [inline modules/axfs/src/root.rs]: Move these 4 `mount_*fs()` functions into a new file `mounts.rs`, and rename to `mounts::devfs()`, `mounts::ramfs()`, ...

File: modules/axfs/src/root.rs [hunk 20]
```diff
@@ -131,6 +131,92 @@ impl VfsNodeOps for RootDirectory {
...
+    let procfs = fs::ramfs::RamFileSystem::new();
+    let proc_root = procfs.root_dir();
+
+    // Create /proc/sys/net/core/somaxconn
+    proc_root.create("sys", VfsNodeType::Dir)?;
+    proc_root.create("sys/net", VfsNodeType::Dir)?;
+    proc_root.create("sys/net/core", VfsNodeType::Dir)?;
+    proc_root.create("sys/net/core/somaxconn", VfsNodeType::File)?;
+    let file_somaxconn = proc_root.clone().lookup("./sys/net/core/somaxconn")?;
+    file_somaxconn.write_at(0, b"4096")?;
```
Comment (equation314) [inline modules/axfs/src/root.rs]: `\n`

File: modules/axfs/src/root.rs [hunk 21]
```diff
@@ -131,6 +131,92 @@ impl VfsNodeOps for RootDirectory {
...
+
+    // Create /proc/sys/vm/overcommit_memory
+    proc_root.create("sys/vm", VfsNodeType::Dir)?;
+    proc_root.create("sys/vm/overcommit_memory", VfsNodeType::File)?;
+    let file_over = proc_root.clone().lookup("./sys/vm/overcommit_memory")?;
+    file_over.write_at(0, b"0")?;
```
Comment (equation314) [inline modules/axfs/src/root.rs]: `\n`

File: modules/axfs/src/root.rs [hunk 22]
```diff
@@ -131,6 +131,92 @@ impl VfsNodeOps for RootDirectory {
             }
         })
     }
+
+    fn rename(&self, src_path: &str, dst_path: &str) -> VfsResult {
+        self.lookup_mounted_fs(src_path, |fs, rest_path| {
+            if rest_path.is_empty() {
+                ax_err!(NotFound)
```
Comment (equation314) [inline modules/axfs/src/root.rs]:
```rust
ax_err!(PermissionDenied) // cannot rename mount points
```

File: ulib/axlibc/c/pthread.c [hunk 23]
```diff
@@ -37,6 +39,20 @@ int pthread_setcanceltype(int new, int *old)
...
+// TODO
+int pthread_cancel(pthread_t t)
+{
+    unimplemented();
+    return 0;
+}
+
 int pthread_mutex_init(pthread_mutex_t *restrict m, const pthread_mutexattr_t *restrict a)
 {
     ax_pthread_mutex_init(m, a);
```
Comment (equation314) [inline ulib/axlibc/c/pthread.c]: `pthread_mutex_*` should return the values return by `ax_pthread_mutex_*`.

**Review by equation314 — APPROVED** [review]

**Pull request merged**